### PR TITLE
feat(SIP-95): new endpoint for extra table metadata

### PR DESCRIPTION
**Pull request by betodealmeida**
### SUMMARY

I started working on [SIP-95](https://github.com/apache/superset/issues/22862), which introduces catalogs to Superset. The introduction of a new hierarchical level, catalog → schema → table, means that some API endpoints need to be updated to support a new parameter.

One of these API endpoints is used to fetch extra metadata about a given table, and the route looks like this:

`/<int:pk>/table_extra/<path:table_name>/<schema_name>/`

There are a few problems with the current route:

- `table_name` is a `path`, because table names can contain slashes.
- `schema_name` is often passed as `null` or `undefined`, so we need special logic to convert it to `None` in the backend.
- `schema_name` should never contain slashes, otherwise the parsing will be wrong.

Because of this, instead of enhancing the existing endpoint to add support for catalogs, I decided to create a new one:

`/<int:pk>/table_metadata/extra/?table=t[&schema=s[&catalog=c]]`

In this new endpoint, `table`, `schema`, and `catalog` are all passed as query parameters. This allows slashes in all of them, and prevents polluting logs with confusing URLs such as `/api/v1/databses/1/table_extra/my_table/undefined/`. Only `table` is required, the other two are optional.

The API calls a new method `get_extra_table_metadata` in the corresponding DB engine spec. The new method takes a `Table` object, which contains a name and optional schema/catalog. To prevent breaking any 3rd-party DB engine specs the new method will issue a warning and call `extra_table_metadata` if it exists and no catalog was specified.

Note that while the old endpoint is not removed in this PR, to prevent breaking changes, the frontend code was updated to use the new endpoint.

In the future, we should remove the old endpoint and also the `extra_table_metadata` method. @michael-s-molina, what do you think of this approach of introducing a new API? Should we mark it as deprecated now and remove it in the next release, or should we mark is as deprecated in 5.0 and remove it in 6.0?

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

N/A

### TESTING INSTRUCTIONS

I updated existing tests and added new unit tests.

### ADDITIONAL INFORMATION

- [X] Has associated issue: https://github.com/apache/superset/issues/22862
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [X] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by michael-s-molina** [comment]
> Should we mark it as deprecated now and remove it in the next release, or should we mark is as deprecated in 5.0 and remove it in 6.0?

You can mark it as deprecated now. Please add a proposal [here](https://github.com/orgs/apache/projects/345) to remove the old endpoint in 5.0. 

> In this new endpoint, table, schema, and catalog are all passed as query parameters. This allows slashes in all of them, and prevents polluting logs with confusing URLs such as /api/v1/databses/1/table_extra/my_table/undefined/. 

Nesting resources or not is a architectural decision, with [trade-offs](https://www.moesif.com/blog/technical/api-design/REST-API-Design-Best-Practices-for-Sub-and-Nested-Resources/), that we need to make consistently in our API. Because we built the API one endpoint at a time, we don't have consistency in nomenclature, response types, params, etc. I'm hoping to fix all these problem with v2 where we would analyze all existing endpoints and propose a consistent API in a SIP. For that reason, any pattern you choose now is acceptable and will be reviewed in v2.

**Comment by betodealmeida** [comment]
Thanks, @michael-s-molina!

> Nesting resources or not is an architectural decision, with [trade-offs](https://www.moesif.com/blog/technical/api-design/REST-API-Design-Best-Practices-for-Sub-and-Nested-Resources/), that we need to make consistently in our API. Because we built the API one endpoint at a time, we don't have consistency in nomenclature, response types, params, etc. I'm hoping to fix all these problems with v2 where we would analyze all existing endpoints and propose a consistent API in a SIP. For that reason, any pattern you choose now is acceptable and will be reviewed in v2.

Oh, I'm not against nesting resources in general. The problem here is that table names can contain slashes, and I think we should assume that schemas and catalogs can too. Imagine we have this route:

`/some/api/<path:catalog>/<path:schema>/<path:table>/`

And we have a table with slashes in the catalog, schema, and name, which we escape with `%2F` when doing a request:

`/some/api/my%2Fcatalog/my%2Fschema/my%2Ftable/`

What the Flask router sees in the request is:

`/some/api/my/catalog/my/schema/my/table/`

Which gets mapped to:

- catalog: `my`
- schema: `catalog`
- table: `my/schema/my/table`

So we'd have to **double escape** the slashes:

`/some/api/my%252Fcatalog/my%252Fschema/my%252Ftable/`

And the route would have unescape them back. For cases like this I think it's easier to pass the parameters as query arguments:

`/some/api/?table=my/table&schema=my/schema&catalog=my/catalog`

While longer, I think it's easier to read. And when some of those parameters are missing it gets better:

`/some/api/?table=my/table`

vs.

`/some/api/undefined/undefined/my%252Ftable/`

So something to think about when designing v2.

**Comment by michael-s-molina** [comment]
> Oh, I'm not against nesting resources in general. The problem here is that table names can contain slashes, and I think we should assume that schemas and catalogs can too.
> So something to think about when designing v2.

Yep! That's indeed a requirement we should consider when designing v2.

File: superset/databases/api.py [hunk 1]
```diff
@@ -812,13 +831,91 @@ def table_extra_metadata(
...
+              $ref: '#/components/responses/401'
+            404:
+              $ref: '#/components/responses/404'
+            500:
+              $ref: '#/components/responses/500'
+        """
+        self.incr_stats("init", self.table_extra_metadata.__name__)
+
+        database = DatabaseDAO.find_by_id(pk)
+        if database is None:
```
Comment (john-bodley) [inline superset/databases/api.py:902]:
```suggestion
        if not (database := DatabaseDAO.find_by_id(pk)):
```

File: superset/databases/schemas.py [hunk 2]
```diff
@@ -1195,3 +1195,27 @@ class OAuth2ProviderResponseSchema(Schema):
...
+
+class QualifiedTableSchema(Schema):
+    """
+    Schema for a qualified table reference.
+
+    Catalog and schema can be ommited, to fallback to default values. Table name must be
+    present.
+    """
+
+    table = fields.String(
```
Comment (john-bodley) [inline superset/databases/schemas.py:1208]: I know I used the term `table` in the `Table` class, but I wonder if this should be name given it's referring to the name of the table. This is also consistent with your description on line 1210.

File: superset/db_engine_specs/base.py [hunk 3]
```diff
@@ -1034,21 +1035,33 @@ def normalize_indexes(cls, indexes: list[dict[str, Any]]) -> list[dict[str, Any]
         return indexes
 
     @classmethod
-    def extra_table_metadata(  # pylint: disable=unused-argument
+    def get_extra_table_metadata(  # pylint: disable=unused-argument
         cls,
         database: Database,
-        table_name: str,
-        schema_name: str | None,
+        table: Table,
```
Comment (john-bodley) [inline superset/db_engine_specs/base.py:1041]: Nice!

File: superset/db_engine_specs/presto.py [hunk 4]
```diff
@@ -1224,14 +1225,20 @@ def expand_data(  # pylint: disable=too-many-locals
...
+        database: Database,
+        table: Table,
     ) -> dict[str, Any]:
         metadata = {}
 
-        if indexes := database.get_indexes(table_name, schema_name):
+        if indexes := database.get_indexes(table.table, table.schema):
             col_names, latest_parts = cls.latest_partition(
-                table_name, schema_name, database, show_first=True, indexes=indexes
+                table.table,
```
Comment (john-bodley) [inline superset/db_engine_specs/presto.py:1237]: Should we be passing a `Table` instance everywhere?

Comment (betodealmeida) [reply superset/db_engine_specs/presto.py:1237]: We should! I'm doing that in the next PR! 😊

File: superset/databases/api.py (hunk 1 above)
Comment (betodealmeida) [reply superset/databases/api.py:902]: Ah, thanks!

File: superset/databases/schemas.py (hunk 2 above)
Comment (betodealmeida) [reply superset/databases/schemas.py:1208]: Yeah, I think the current terminology is confusing, there's a lot of `table.table` in the codebase. Let me change it.

File: superset/databases/api.py [hunk 5]
```diff
@@ -812,13 +831,92 @@ def table_extra_metadata(
...
+        try:
+            parameters = QualifiedTableSchema().load(request.args)
+        except ValidationError as ex:
+            raise InvalidPayloadSchemaError(ex) from ex
+
+        table = Table(parameters["name"], parameters["schema"], parameters["catalog"])
+        try:
+            security_manager.raise_for_access(database=database, table=table)
+        except SupersetSecurityException as ex:
+            # instead of raising 403, raise 404 to hide table existence
```
Comment (dpgaspar) [inline superset/databases/api.py:915]: nice!

File: superset/db_engine_specs/base.py [hunk 6]
```diff
@@ -1034,21 +1035,33 @@ def normalize_indexes(cls, indexes: list[dict[str, Any]]) -> list[dict[str, Any]
...
 
         :param database: Database instance
-        :param table_name: Table name
-        :param schema_name: Schema name
+        :param table: A Table instance
         :return: Engine-specific table metadata
         """
-        # TODO: Fix circular import caused by importing Database
+        # old method that doesn't work with catalogs
+        if hasattr(cls, "extra_table_metadata"):
```
Comment (dpgaspar) [inline superset/db_engine_specs/base.py:1051]: optional: 
do:
``` python
if not hasattr(cls, "extra_table_metadata"):
    return {}
...
```
to reduce nesting

**Review by dpgaspar — APPROVED** [review]

File: superset/db_engine_specs/base.py (hunk 6 above)
Comment (betodealmeida) [reply superset/db_engine_specs/base.py:1051]: Nice, I'll fix it in the next PR!

**Pull request merged**